File: Year02/2nd semester/MPP/Lab/mpp-proiect-csharp-AlexandruPaul21/CsharpProj/repository/FlightRepository.cs
using System.Configuration;
using CsharpProj.model;
using log4net;
using Npgsql;

namespace CsharpProj.repository;

public class FlightRepository : IFlightRepository
{

    //private readonly Logger logger = new Logger();
    
    private static readonly ILog logger = LogManager.GetLogger("FlightRepository");
    
    public Flight findOne(long Id)
    {
        logger.Info("Flight - findOne - Enter");
        Flight flight = null!;

        var sql = "SELECT * FROM flights WHERE id = @id";

        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                cmd.Parameters.AddWithValue("@id", Id);
                var reader = cmd.ExecuteReader();
                if (reader.Read())
                {
                    long id = reader.GetInt32(0);
                    var destination = reader.GetString(1);
                    var departureDateTime = reader.GetDateTime(2);
                    var airport = reader.GetString(3);
                    var availableSeats = reader.GetInt32(4);

                    flight = new Flight(destination, departureDateTime, airport, availableSeats);
                    flight.Id = id;
                }
            }
        }
        logger.Info("Flight - findOne - Exit");
        return flight;
    }

    public IEnumerable<Flight> findAll()
    {
        logger.Info("Flight - findAll - Enter");
        var flights = new List<Flight>();

        var sql = "SELECT * FROM flights";

        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                var reader = cmd.ExecuteReader();
                while (reader.Read())
                {
                    long id = reader.GetInt32(0);
                    var destination = reader.GetString(1);
                    var departureDateTime = reader.GetDateTime(2);
                    var airport = reader.GetString(3);
                    var availableSeats = reader.GetInt32(4);

                    Flight flight = new Flight(destination, departureDateTime, airport, availableSeats);
                    flight.Id = id;
                    
                    flights.Add(flight);
                }
            }
        }
        logger.Info("Flight - findAll - Exit");
        return flights;
    }

    public Flight save(Flight entity)
    {
        logger.Info("Flight - save - Enter");
        var sql = "INSERT INTO flights VALUES (@id, @destination, @departure, @airport, @available)";

        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                cmd.Parameters.AddWithValue("@id", entity.Id);
                cmd.Parameters.AddWithValue("@destination", entity.Destination);
                cmd.Parameters.AddWithValue("@departure", entity.DepartureDateTime);
                cmd.Parameters.AddWithValue("@airport", entity.Airport);
                cmd.Parameters.AddWithValue("@available", entity.AvailableSeats);

                var rowsAff = cmd.ExecuteNonQuery();

                if (rowsAff != 0)
                {
                    entity = null!;
                }
            }
        }
        logger.Info("Flight - save - Exit");
        return entity;
    }

    public IEnumerable<Flight> findFlightByDestinationAndDate(string destination, DateTime dateTime)
    {
        DateTime start = new DateTime(dateTime.Year, dateTime.Month, dateTime.Day, 0, 0, 0);
        DateTime end = new DateTime(dateTime.Year, dateTime.Month, dateTime.Day, 23, 59, 59);
        logger.Info("Flight - findFlightByDestinationAndDate - Enter");
        
        var flights = new List<Flight>();
        
        var sql = "SELECT * FROM flights WHERE destination = @destination AND departure_date_time BETWEEN @start AND @end";

        using (var connection = new NpgsqlConnection(ConfigurationManager.ConnectionStrings["Test"].ConnectionString))
        {
            connection.Open();
            using (var cmd = new NpgsqlCommand(sql, connection))
            {
                cmd.Parameters.AddWithValue("@destination", destination);
                cmd.Parameters.AddWithValue("@start", start);
                cmd.Parameters.AddWithValue("@end", end);
                var reader = cmd.ExecuteReader();
                while (reader.Read())
                {
                    long id = reader.GetInt32(0);
                    var dest = reader.GetString(1);
                    var departureDateTime = reader.GetDateTime(2);
                    var airport = reader.GetString(3);
                    var availableSeats = reader.GetInt32(4);

                    Flight flight = new Flight(dest, departureDateTime, airport, availableSeats);
                    flight.Id = id;
                    
                    flights.Add(flight);
                }
            }
        }
        logger.Info("Flight - findFlightByDestinationAndDate - Exit");
        return flights;
    }
}

File: Year02/2nd semester/MPP/Lab/mpp-proiect-java-AlexandruPaul21/JavaRest/src/main/java/org/example/persistance/jdbc/FlightRepository.java
package org.example.persistance.jdbc;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.Flight;
import org.example.repository.IFlightRepository;

import java.sql.*;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class FlightRepository implements IFlightRepository {

    private final JdbcUtils jdbcUtils;
    private static final Logger logger = LogManager.getLogger();

    public FlightRepository(Properties properties) {
        logger.info("Initialising FlightRepository with properties: {}", properties);
        jdbcUtils = new JdbcUtils(properties);
    }

    @Override
    public Flight findOne(Long aLong) {
        logger.traceEntry();
        String sql = "SELECT * FROM flights WHERE id = ?";
        Flight flight = null;

        try (
                Connection connection = jdbcUtils.getConnection()
                ) {
            PreparedStatement preparedStatement = connection.prepareStatement(sql);
            preparedStatement.setLong(1, aLong);

            ResultSet resultSet = preparedStatement.executeQuery();

            if (resultSet.next()) {
                Long id = resultSet.getLong("id");
                String destination = resultSet.getString("destination");
                LocalDateTime departureDateTime = resultSet
                        .getTimestamp("departure_date_time").toLocalDateTime();
                String airport = resultSet.getString("airport");
                int availableSeats = resultSet.getInt("available_seats");

                flight = new Flight(destination, departureDateTime, airport, availableSeats);
                flight.setId(id);
            }
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }

        logger.traceExit();
        return flight;
    }

    @Override
    public Iterable<Flight> findAll() {
        logger.traceEntry();
        String sql = "SELECT * FROM flights";

        List<Flight> flights = new ArrayList<>();

        try (
                Connection connection = jdbcUtils.getConnection()
                ) {

            PreparedStatement preparedStatement = connection.prepareStatement(sql);

            ResultSet resultSet = preparedStatement.executeQuery();

            while (resultSet.next()) {
                Long id = resultSet.getLong("id");
                String destination = resultSet.getString("destination");
                LocalDateTime departureDateTime = resultSet
                        .getTimestamp("departure_date_time").toLocalDateTime();
                String airport = resultSet.getString("airport");
                int availableSeats = resultSet.getInt("available_seats");

                Flight flight = new Flight(destination, departureDateTime, airport, availableSeats);
                flight.setId(id);

                flights.add(flight);
            }
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }
        logger.traceExit();
        return flights;
    }

    @Override
    public Flight save(Flight entity) {
        logger.traceEntry();

        String sql = "INSERT INTO flights VALUES (?, ?, ?, ?, ?)";

        try (
                Connection connection = jdbcUtils.getConnection()
                ) {
            PreparedStatement preparedStatement = connection.prepareStatement(sql);
            preparedStatement.setLong(1, entity.getId());
            preparedStatement.setString(2, entity.getDestination());
            preparedStatement.setTimestamp(3, Timestamp.valueOf(entity.getDepartureDateTime()));
            preparedStatement.setString(4, entity.getAirport());
            preparedStatement.setInt(5, entity.getAvailableSeats());

            preparedStatement.executeUpdate();
            entity = null;
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }
        logger.traceExit();
        return entity;
    }

    @Override
    public Iterable<Flight> findFlightByDestinationAndDate(String destination, LocalDateTime dateTime) {
        logger.traceEntry("findFlightByDestinationAndDate: destination = {}, dateTime = {}", destination, dateTime);
        LocalDateTime start = LocalDateTime.of(dateTime.getYear(), dateTime.getMonth(), dateTime.getDayOfMonth(), 0, 0);
        LocalDateTime end = LocalDateTime.of(dateTime.getYear(), dateTime.getMonth(), dateTime.getDayOfMonth(), 23, 59);
        String sql = "SELECT * FROM flights WHERE destination = ? AND departure_date_time BETWEEN ? AND ?";

        try (
                Connection connection = jdbcUtils.getConnection();
                PreparedStatement preparedStatement = connection.prepareStatement(sql)
                ) {
            preparedStatement.setString(1, destination);
            preparedStatement.setTimestamp(2, Timestamp.valueOf(start));
            preparedStatement.setTimestamp(3, Timestamp.valueOf(end));

            ResultSet resultSet = preparedStatement.executeQuery();

            List<Flight> flights = new ArrayList<>();

            while (resultSet.next()) {
                Long id = resultSet.getLong("id");
                String destination1 = resultSet.getString("destination");
                LocalDateTime departureDateTime = resultSet
                        .getTimestamp("departure_date_time").toLocalDateTime();
                String airport = resultSet.getString("airport");
                int availableSeats = resultSet.getInt("available_seats");

                Flight flight = new Flight(destination1, departureDateTime, airport, availableSeats);
                flight.setId(id);

                flights.add(flight);
            }
            logger.traceExit();
            return flights;
        } catch (SQLException e) {
            logger.error(e);
            e.printStackTrace();
        }
        logger.traceExit();
        return null;
    }
}
